util/etcd/etcd.go

```diff
@@ -0,0 +1,382 @@
+	IsLearner bool
+	// Alarms is the list of alarms for a member
+	Alarms []AlarmType
+}
+
+type Status struct {
+	// Leader is the member ID which the responding member believes is the current leader.
+	Leader uint64
+
+	// Version is the cluster protocol version used by the responding member
+	Version string
+
+	// Errors contains alarm/health information and status.
+	Errors []string
+
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+
+	// RaftIndex is the current raft committed index of the responding member.
+	RaftIndex uint64
+
+	// RaftTerm is the current raft term of the responding member.
+	RaftTerm uint64
+
+	// RaftAppliedIndex is the current raft applied index of the responding member.
+	RaftAppliedIndex uint64
+
+	// DbSizeInUse is the size of the backend database logically in use, in bytes, of the responding member.
+	DbSizeInUse int64
+
+	// DdSize is the size of the backend database physically allocated, in bytes, of the responding member.
+	DbSize int64
+}
+
+// New creates a new etcd client with a custom dialer
+func New(dialer grpcDial, endpoint string, tlsConfig *tls.Config, logger logr.Logger) (*Client, error) {
+	if endpoint == "" {
+		endpoint = "localhost"
+	}
+	c := &Client{
+		dialerContext: dialer,
+		tlsConfig:     tlsConfig,
+		timeout:       etcdTimeout,
+		backoffParams: etcdBackoff,
+		endpoint:      endpoint,
+		log:           logger,
+	}
+	return c, nil
+}
+
+func (c *Client) newEtcdClient() (*clientv3.Client, error) {
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{c.endpoint},
+		DialTimeout: etcdTimeout,
+		DialOptions: []grpc.DialOption{
+			grpc.WithBlock(), // block until the underlying connection is up
+			grpc.WithContextDialer(c.dialerContext),
+		},
+		TLS: c.tlsConfig,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return cli, nil
+}
+
+// Members retrieves a list of etcd members
+func (c *Client) Members() (*[]Member, error) {
+	cli, err := c.newEtcdClient()
+	if err != nil {
+		return nil, err
+	}
+	defer cli.Close()
+	var lastError error
+	var resp *clientv3.MemberListResponse
+	err = wait.ExponentialBackoff(c.backoffParams, func() (bool, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+		resp, err = cli.MemberList(ctx)
+		cancel()
+		if err == nil {
+			return true, nil
+		}
+		c.log.Error(err, "failed to get etcd member list")
+		lastError = err
+		return false, nil
+	})
+	if err != nil {
+		return nil, lastError
+	}
+
+	alarms, err := c.Alarms()
+	if err != nil {
+		return nil, err
+	}
+
+	members := make([]Member, len(resp.Members))
+	for i, m := range resp.Members {
+		newMember := clientMemberToclientMember(m)
+		for _, c := range *alarms {
+			if c.MemberID == newMember.ID {
+				newMember.Alarms = append(newMember.Alarms, c.Type)
+			}
+			members[i] = newMember
+		}
+	}
+
+	return &members, nil
+}
+
+func clientMemberToclientMember(m *etcdserverpb.Member) Member {
```

Comment:
Should this be called clientMemberToMember or something mentioning protobuf/encoded data?

Reply:
yes. this is find and replace gone wrong.

Reply:
:joy: For a second I thought my eyes were playing tricks on me...